bgpd/bgp_bmp.c

```diff
@@ -271,29 +312,43 @@ static void bmp_per_peer_hdr(struct stream *s, struct peer *peer,
 	stream_putc(s, flags);
 
 	/* Peer Distinguisher */
-	memset (&peer_distinguisher[0], 0, 8);
-	stream_put(s, &peer_distinguisher[0], 8);
+	stream_put(s, (uint8_t *)&peer_distinguisher, 8);
 
 	/* Peer Address */
-	if (peer->su.sa.sa_family == AF_INET6)
-		stream_put(s, &peer->su.sin6.sin6_addr, 16);
-	else if (peer->su.sa.sa_family == AF_INET) {
+	/* Set to 0 if it's a LOC-RIB INSTANCE (RFC 9069) or if it's not an
+	 * IPv4/6 address
+	 */
+	if (is_locrib || (peer->su.sa.sa_family != AF_INET6 &&
+			  peer->su.sa.sa_family != AF_INET)) {
 		stream_putl(s, 0);
 		stream_putl(s, 0);
 		stream_putl(s, 0);
-		stream_put_in_addr(s, &peer->su.sin.sin_addr);
-	} else {
 		stream_putl(s, 0);
+	} else if (peer->su.sa.sa_family == AF_INET6)
+		stream_put(s, &peer->su.sin6.sin6_addr, 16);
+	else if (peer->su.sa.sa_family == AF_INET) {
 		stream_putl(s, 0);
 		stream_putl(s, 0);
 		stream_putl(s, 0);
+		stream_put_in_addr(s, &peer->su.sin.sin_addr);
 	}
 
 	/* Peer AS */
-	stream_putl(s, peer->as);
+	/* set peer ASN but for LOC-RIB INSTANCE (RFC 9069) put the local bgp
+	 * ASN if available or 0
+	 */
+	as_t asn = !is_locrib ? peer->as : bgp ? bgp->as : 0L;
+
+	stream_putl(s, asn);
 
 	/* Peer BGP ID */
-	stream_put_in_addr(s, &peer->remote_id);
+	/* set router-id but for LOC-RIB INSTANCE (RFC 9069) put the instance
+	 * router-id if available or 0
+	 */
+	struct in_addr *bgp_id =
+		!is_locrib ? &peer->remote_id : bgp ? &bgp->router_id : NULL;
```

Comment:
I think this should be `local_id`, not `remote_id`. 

```
Set the ID to the router-id of the VRF instance if VRF is used; otherwise, set to the global instance router-id.
```

Reply:
In case of `!is_locrib` aka *adj-rib-in*:
[RFC7854 BMP (Adj-In-Pre/Post)](https://datatracker.ietf.org/doc/html/rfc7854#section-4.2)
  ```
 o  Peer BGP ID: The BGP Identifier of the peer from which the
      encapsulated PDU was received.
```
I am not sure what the difference between `local_id` and `remote_id` is. Could you explain? 
I chose `remote_id` because the original bmp code used it too.


In case of `is_locrib` :
[RFC9069 BMP Local-RIB](https://datatracker.ietf.org/doc/html/rfc9069#section-5.1)
```
   Peer BGP ID:  Set the ID to the router-id of the VRF instance if VRF
      is used; otherwise, set to the global instance router-id.
```
so i would say the `bgp->router_id` part is correct right?